### PR TITLE
Support rollback nodes in PostCommitValidation

### PR DESCRIPTION
**Pull request by cocreature**
With that and the other in-flight PRs, #9400 now passes against Daml
on SQL.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala [hunk 1]
```diff
@@ -122,29 +122,23 @@ private[appendonlydao] object PostCommitValidation {
...
-          case ((created, ids), (_, e: Exercise)) if !divulged(e.targetCoid) =>
-            (created, ids + e.targetCoid)
-          case ((created, ids), (_, f: Fetch)) if !divulged(f.coid) =>
-            (created, ids + f.coid)
-          case ((created, ids), (_, l: LookupByKey)) =>
-            (created, l.result.filterNot(divulged).fold(ids)(ids + _))
-          case ((created, ids), _) => (created, ids)
-        }
-      referred.diff(createdInTransaction)
+      transaction.inputContracts.diff(divulged)
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala:125]: If this is a one-liner used only in one place I'd suggest to just inline it.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala [hunk 2]
```diff
@@ -170,18 +164,49 @@ private[appendonlydao] object PostCommitValidation {
...
+    * @param contracts Active contracts created in
+    *  the current transaction that have a key indexed
+    *  by a hash of their key.
+    * @param removed Hashes of contract keys that are known to
+    *  to be archived. Note that a later create with the same
+    *  key will remove the entry again.
+    */
+  private final case class ActiveState(
+      val contracts: Map[Hash, ContractId],
+      val removed: Set[Hash],
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala:180]:
```suggestion
      contracts: Map[Hash, ContractId],
      removed: Set[Hash],
```
`val` is implicit for all fields of a `case class`.

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala [hunk 3]
```diff
@@ -204,29 +229,40 @@ private[appendonlydao] object PostCommitValidation {
...
+    def beginRollback(): State =
       copy(
-        contracts = contracts.updated(key.hash, id),
-        removed = removed - key.hash,
+        rollbackStack = currentState :: rollbackStack
       )
 
+    def endRollback(): State = rollbackStack match {
+      case Nil =>
+        throw new IllegalStateException(s"Internal error: rollback end but rollbackStack was empty")
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala:239]:
```suggestion
        throw new IllegalStateException("Internal error: rollback end but rollbackStack was empty")
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala [hunk 4]
```diff
@@ -122,29 +122,23 @@ private[dao] object PostCommitValidation {
...
-          case ((created, ids), (_, e: Exercise)) if !divulged(e.targetCoid) =>
-            (created, ids + e.targetCoid)
-          case ((created, ids), (_, f: Fetch)) if !divulged(f.coid) =>
-            (created, ids + f.coid)
-          case ((created, ids), (_, l: LookupByKey)) =>
-            (created, l.result.filterNot(divulged).fold(ids)(ids + _))
-          case ((created, ids), _) => (created, ids)
-        }
-      referred.diff(createdInTransaction)
+      transaction.inputContracts.diff(divulged)
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:125]: Same as above, I'd recommend to just inline this.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala [hunk 5]
```diff
@@ -170,18 +164,49 @@ private[dao] object PostCommitValidation {
...
+    * @param contracts Active contracts created in
+    *  the current transaction that have a key indexed
+    *  by a hash of their key.
+    * @param removed Hashes of contract keys that are known to
+    *  to be archived. Note that a later create with the same
+    *  key will remove the entry again.
+    */
+  private final case class ActiveState(
+      val contracts: Map[Hash, ContractId],
+      val removed: Set[Hash],
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:180]:
```suggestion
      contracts: Map[Hash, ContractId],
      removed: Set[Hash],
```

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala [hunk 6]
```diff
@@ -204,29 +229,40 @@ private[dao] object PostCommitValidation {
...
+    def beginRollback(): State =
       copy(
-        contracts = contracts.updated(key.hash, id),
-        removed = removed - key.hash,
+        rollbackStack = currentState :: rollbackStack
       )
 
+    def endRollback(): State = rollbackStack match {
+      case Nil =>
+        throw new IllegalStateException(s"Internal error: rollback end but rollbackStack was empty")
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/PostCommitValidation.scala:239]:
```suggestion
        throw new IllegalStateException("Internal error: rollback end but rollbackStack was empty")
```

**Review by stefanobaghino-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala (hunk 1 above)
Comment (cocreature) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala:125]: I considered it but it seemed easier for review at least to only replace the implementation but leave it in a separate definition. Don’t feel strongly though so happy to inline it.

Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/PostCommitValidation.scala:125]: Good call as a first step. 🙇🏻

**Review by nickchapman-da — APPROVED** [review]
LGTM

**Pull request merged**